### PR TITLE
[Core] fix _get_num_new_tokens() for _schedule_default()

### PR DESCRIPTION
**Pull request by Yuyi-Ao**
If chunked-prefill is not enabled, the vLLM scheduler calls _schedule_default(). Within this function, the _schedule_running() method is called, where it is possible that the scheduler may attempt to schedule more tokens than the allocated token budget, leading to an assertion failure:
```
        assert (budget.num_batched_tokens <=
                self.scheduler_config.max_num_batched_tokens)
```
The scheduling policy should ensure the scheduler does not schedule more tokens than the token budget. This assertion failure occurs when decoding requests exceeds token_budget. One can easily trigger it by setting self.scheduler_config.max_num_batched_tokens to a smaller value, such as 200 or sending lots of requests at one time. The conditional statement added only takes care of decoding tokens in _schedule_default().

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by George-ao** [comment]
Hi @DarkLight1337, can you take a look at this change?

**Comment by DarkLight1337** [comment]
Sorry I'm not really familiar with this part of the code. See if others can help.

File: vllm/core/scheduler.py [hunk 1]
```diff
@@ -1333,4 +1333,7 @@ def _get_num_new_tokens(self, seq_group: SequenceGroup,
         if enable_chunking and len(seqs) == 1:
             num_new_tokens = min(num_new_tokens,
                                  budget.remaining_token_budget())
+        if not enable_chunking and status == SequenceStatus.RUNNING:
+            num_new_tokens = min(num_new_tokens,
+                                 budget.remaining_token_budget())
```
Comment (cadedaniel) [inline vllm/core/scheduler.py:1338]: I think this has a bug if the sequence is running with beam search (`len(seqs)` > 1 and status=running --> number of scheduled decode tokens is inconsistent with budget).

**Review by cadedaniel — COMMENTED** [review]
Can we add some tests for this case? Can be a unit test or an e2e test like this that hits this case https://github.com/vllm-project/vllm/blob/24154f8618aa5da695b18cbbd71704f59803ec6b/tests/core/block/e2e/test_correctness.py#L297

**Comment by cadedaniel** [comment]
also cc @rkooo567

**Review by rkooo567 — COMMENTED** [review]
can you add a unit test? are you saying if num_max_batched_tokens < number of decode requests, this happens?

**Comment by George-ao** [comment]
> can you add a unit test? are you saying if num_max_batched_tokens < number of decode requests, this happens?

Thanks for reply, but I realize I was wrong. It comes to me that this change seems unnecessary as max_num_batched_tokens must be greater than or equal to max_num_seqs.

**Pull request closed without merging**